capsules/extra/src/nmea_i2c.rs

```diff
@@ -0,0 +1,181 @@
+                    Ok(()) => {
+                        self.sentence_buffer.replace(sentence);
+                        Ok(())
+                    }
+                    Err((e, buf)) => {
+                        self.i2c_buffer.replace(buf);
+                        Err((e.into(), sentence))
+                    }
+                }
+            }
+            None => Err((ErrorCode::BUSY, sentence)),
+        }
+    }
+}
+
+impl<'a, I: I2CDevice> I2CClient for I2cNmea<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        let nmea_buf = self.sentence_buffer.take().unwrap();
+        let i2c_buf_len = buffer.len();
+
+        if let Err(e) = status {
+            self.i2c_buffer.replace(buffer);
+
+            self.client.map(|call| {
+                call.callback(nmea_buf, 0, Err(e.into()));
+            });
+
+            return;
+        }
+
+        let string = match str::from_utf8(buffer) {
+            Ok(utf8) => utf8,
+            Err(_e) => {
+                self.i2c_buffer.replace(buffer);
+
+                self.client.map(|call| {
+                    call.callback(nmea_buf, 0, Err(ErrorCode::NOSUPPORT));
+                });
+                return;
+            }
+        };
+        let mut nmea_offset = self.nmea_offset.get();
+
+        if let Some(location) = string.find('$') {
+            // This is the start of a new sentence
+
+            if (nmea_offset + location) > nmea_buf.len() {
+                // We will overflow our buffer, just drop the packet and try again
+                self.nmea_offset.set(0);
+                self.sentence_buffer.replace(nmea_buf);
+
+                if let Err((e, buf)) = self.i2c.read(buffer, i2c_buf_len) {
+                    self.i2c_buffer.replace(buf);
+
+                    self.client.map(|call| {
+                        call.callback(self.sentence_buffer.take().unwrap(), 0, Err(e.into()));
+                    });
+                }
+                return;
+            }
+
+            // First copy the remainder of the sentence to the buffer
+            nmea_buf[nmea_offset..(nmea_offset + location)].copy_from_slice(&buffer[0..location]);
+            nmea_offset += location;
+
+            // Now parse the sentence
+            let sentence = match str::from_utf8(&nmea_buf[0..nmea_offset]) {
+                Ok(utf8) => utf8,
+                Err(_e) => {
+                    self.i2c_buffer.replace(buffer);
+
+                    self.client.map(|call| {
+                        call.callback(nmea_buf, 0, Err(ErrorCode::NOSUPPORT));
+                    });
+                    return;
+                }
+            };
+
+            if sentence.starts_with('$') {
+                // At this point we have a sentence with a `$` at the start.
+                // We report it back to the caller.
+                // We loose the rest of the data we just read though
```

Comment:
This would be more robust and more straightforward to just have this file own its own buffer that it can read messages into, and then when it finds two `$` characters it copies the string to `self.sentence_buffer`, triggers the callback, moves the remaining bytes to the beginning of the new buffer, and continues receiving.

Right now unless we get lucky with the message lengths we are dropping sentences, and if any utf-8 parsing errors happen we can drop even more messages.

Reply:
In practice you receive all of the NMEA messages. They are just continuously printed over and over again. A user expects to received a single sentence. There is no control over what that sentence is.

If a user wants more control they can use https://github.com/tock/tock/pull/3867 to directly communicate with the GNSS device.

> This would be more robust and more straightforward to just have this file own its own buffer that it can read messages into, and then when it finds two `$` characters it copies the string to `self.sentence_buffer`, triggers the callback, moves the remaining bytes to the beginning of the new buffer, and continues receiving.

I don't think you understand how quickly the data is being pushed out from the device. Unless this buffer is **really** big then we have the same situation where multiple sentences will be replaced in the time it takes for the callback to trigger and the caller to recall the `read` command.

> Right now unless we get lucky with the message lengths we are dropping sentences, and if any utf-8 parsing errors happen we can drop even more messages.

No matter what you are doing to drop sentences. But they get printed again very quickly so it doesn't matter

Reply:
My concern is this file isn't named after a particular chip, so it's hard to say this behavior is true of all I2C NMEA devices. The UART NMEA devices I've used print data at 1Hz. This [random device](https://cdn.sparkfun.com/assets/parts/1/2/2/8/0/GTOP_NMEA_over_I2C_Application_Note.pdf) I found seems to also have data once per second:

<img width="852" alt="image" src="https://github.com/tock/tock/assets/1467890/90a54140-a1d5-46dd-97ae-5dca52bc361a">

At that rate, I think a user would expect to get every sentence.

Is the duplicate data actually "new" data from the device, or does your device just return the same sentence over and over on every read, until new data is available, at which point it returns that?

I just don't think we want a generic looking capsule to be tailored to one (unnamed) chip and for a new user to find that it drops half of their messages unintentionally.

Reply:
> Is the duplicate data actually "new" data from the device, or does your device just return the same sentence over and over on every read, until new data is available, at which point it returns that?

It is "new" data, in the sense that it will update if something happens. But it's not as if you miss the important event.

If a user wants to ensure they don't loose any data they can read a byte at a time. That way nothing will be thrown away. The read length is configurable by the buffer length.

We can keep the remainder of the sentence and use it next time, but that risks corrupting the next read if the user waits too long (as the device is now outputting the next sentence).

> I just don't think we want a generic looking capsule to be tailored to one (unnamed) chip and for a new user to find that it drops half of their messages unintentionally.

So you propose instead to create a ~300 byte buffer to store the sentences in (NMEA sentences in theory are a max of 255 bytes)?

Which still results in lost data if the caller isn't quick enough to read the sentence? I don't see how it's any better but does result in a much larger memory overhead.

It's also worth noting that continuously reading the data falls apart when you are interested in power saving. For example it's common to sleep for a minute before requesting a new location. Letting the GNSS device go into low power mode saves a lot of power

Reply:
I think anything low power would use interrupts.

Reply:
@alistair23 what's the chip you're implementing against? Do you have a reference?

Reply:
How would interrupts help?

Isn't your proposal to continuously read the data so we never loose anything? In that case you are just always reading, so there is no need for interrupts.

In my approach interrupts are indeed how you read the data. When an interrupt occurs you can then call `read_sentence()` to get the latest data.